run/batch_applier.go

```diff
@@ -43,91 +50,162 @@ type ApplyOptions struct {
 // Apply takes a list of files and attempts an apply command on each.
 // It returns two lists of ApplyAttempts - one for files that succeeded, and one for files that failed.
 func (a *BatchApplier) Apply(applyList []string, options *ApplyOptions) ([]ApplyAttempt, []ApplyAttempt) {
+	if a.WorkerCount == 0 {
+		a.WorkerCount = defaultBatchApplierWorkerCount
+	}
 	successes := []ApplyAttempt{}
 	failures := []ApplyAttempt{}
+	wg := sync.WaitGroup{}
+	mutex := sync.Mutex{}
+
+	paths := make(chan string, len(applyList))
+
+	for i := 0; i < a.WorkerCount; i++ {
+		wg.Add(1)
+		go func(paths <-chan string) {
+			defer wg.Done()
+			for path := range paths {
+				appliedFile, success := a.apply(path, options)
+				if appliedFile == nil {
+					continue
+				}
+
+				mutex.Lock()
+				if success {
+					successes = append(successes, *appliedFile)
+					log.Logger.Info(fmt.Sprintf("%v\n%v", appliedFile.Command, appliedFile.Output))
+				} else {
+					failures = append(failures, *appliedFile)
+					log.Logger.Warn(fmt.Sprintf("%v\n%v", appliedFile.Command, appliedFile.ErrorMessage))
+				}
+				a.Metrics.UpdateNamespaceSuccess(path, success)
+				mutex.Unlock()
+			}
+		}(paths)
+	}
 
 	for _, path := range applyList {
-		log.Logger.Info(fmt.Sprintf("Applying dir %v", path))
-		ns := filepath.Base(path)
-		kaa, err := a.KubeClient.NamespaceAnnotations(ns)
-		if err != nil {
-			log.Logger.Error("Error while getting namespace annotations, defaulting to kube-applier.io/enabled=false", "error", err)
-			continue
-		}
+		paths <- path
+	}
 
-		enabled, err := strconv.ParseBool(kaa.Enabled)
-		if err != nil {
-			log.Logger.Info("Could not get value for kube-applier.io/enabled", "error", err)
-			continue
-		} else if !enabled {
-			log.Logger.Info("Skipping namespace", "kube-applier.io/enabled", enabled)
-			continue
+	close(paths)
+	wg.Wait()
+
+	sort.Slice(successes, func(i, j int) bool {
+		indexI := -1
+		indexJ := -1
+		found := 0
+		for k, v := range applyList {
+			if successes[i].FilePath == v {
+				indexI = k
+				found++
+			}
+			if successes[j].FilePath == v {
+				indexJ = k
+				found++
+			}
+			if found == 2 {
+				break
+			}
 		}
-
-		dryRun, err := strconv.ParseBool(kaa.DryRun)
-		if err != nil {
-			log.Logger.Info("Could not get value for kube-applier.io/dry-run", "error", err)
-			dryRun = false
+		return indexI < indexJ
+	})
+	sort.Slice(failures, func(i, j int) bool {
+		indexI := -1
+		indexJ := -1
+		found := 0
+		for k, v := range applyList {
+			if failures[i].FilePath == v {
+				indexI = k
+				found++
+			}
+			if failures[j].FilePath == v {
+				indexJ = k
+				found++
+			}
+			if found == 2 {
+				break
+			}
 		}
+		return indexI < indexJ
+	})
```

Comment:
Rather than duplicating the sort logic here, you could define a single method to do this. Something like `sortApplyAttempts`.